### PR TITLE
Tests To recognize the ApartmentStateThreadAttribute on Main  Function

### PR DESCRIPTION
**Pull request by Anipik**
Related PR - https://github.com/dotnet/coreclr/pull/15512

### CODE REVIEW

File: src/System.Threading.Thread/tests/MTAMain/MTAMain.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace VoidMainWithExitCodeApp
+{
+    internal static class Program
+    {
+        [MTAThread]
+        static void Main(string[] args)
+        {
+            PropertyInfo set_ExitCode = typeof(Environment).GetTypeInfo().GetDeclaredProperty("ExitCode");
+            MethodInfo Exit = typeof(Environment).GetTypeInfo().GetDeclaredMethod("Exit");
```
Comment (jkotas) [inline src/System.Threading.Thread/tests/MTAMain/MTAMain.cs]: Why are you using reflection for all this?

File: src/System.Threading.Thread/tests/STAMain/STAMain.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        [STAThread]
+        static int Main(string[] args)
+        {
+            int retValue = 0;
+
+            bool mode = bool.Parse(args[0]);
+            if (mode)
+            {
+                if (Thread.CurrentThread.GetApartmentState().ToString().Equals("STA"))
```
Comment (jkotas) [inline src/System.Threading.Thread/tests/STAMain/STAMain.cs]: This does not need to convert to string

File: src/System.Threading.Thread/tests/ThreadTests.cs [hunk 3]
```diff
@@ -139,6 +144,58 @@ yield return
...
+        }
+
+        [Theory]
+        [InlineData(1)]
+        [InlineData(2)]
+        [InlineData(3)]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+        public static void ApartmentState_NoAtributePresent(int mode)
+        {
+            const string AppName = "UnKnownMain.exe";
```
Comment (jkotas) [inline src/System.Threading.Thread/tests/ThreadTests.cs]: UnKnown is a single world. It should be cased as Unknown

File: src/System.Threading.Thread/tests/Program.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace tests
+{
```
Comment (jkotas) [inline src/System.Threading.Thread/tests/Program.cs]: Is this good for anything?

File: src/System.Threading.Thread/tests/MTAMain/MTAMain.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        [MTAThread]
+        static int Main(string[] args)
+        {
+            int retValue = 0;
+
+            bool mode = bool.Parse(args[0]);
+            if (mode)
+            {
+                if (Thread.CurrentThread.GetApartmentState().Equals(ApartmentState.MTA))
```
Comment (jkotas) [inline src/System.Threading.Thread/tests/MTAMain/MTAMain.cs]: You can just do `...GetApartmentState() == ApartmentState.MTA)`. Equals is just a more verbose and more expensive way to achieve the same.

File: src/System.Threading.Thread/tests/STAMain/STAMain.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            int retValue = 0;
+
+            bool mode = bool.Parse(args[0]);
+            if (mode)
+            {
+                if (Thread.CurrentThread.GetApartmentState().Equals(ApartmentState.STA))
+                {
+                    retValue = 1;
+                }
+
```
Comment (jkotas) [inline src/System.Threading.Thread/tests/STAMain/STAMain.cs]: Nit: Extra line

File: src/System.Threading.Thread/tests/MTAMain/MTAMain.exe.config [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <runtime>
+    <developmentMode developerInstallation="true" />
```
Comment (jkotas) [inline src/System.Threading.Thread/tests/MTAMain/MTAMain.exe.config]: What is this for?

File: src/System.Threading.Thread/tests/STAMain/STAMain.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+                    retValue = 1;
+                }    
+            }
+            else
+            {
+                try
+                {
+                    Thread.CurrentThread.SetApartmentState(ApartmentState.MTA);
+                }
+                catch(InvalidOperationException)
```
Comment (danmosemsft) [inline src/System.Threading.Thread/tests/STAMain/STAMain.cs]: spacing after catch

File: src/System.Threading.Thread/tests/STAMain/STAMain.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        static int Main(string[] args)
+        {
+            int retValue = 0;
+
+            bool mode = bool.Parse(args[0]);
+            if (mode)
+            {
+                if (Thread.CurrentThread.GetApartmentState() == ApartmentState.STA)
+                {
+                    retValue = 1;
```
Comment (danmosemsft) [inline src/System.Threading.Thread/tests/STAMain/STAMain.cs]: Isn't this a success? should return 0

File: src/System.Threading.Thread/tests/MTAMain/MTAMain.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        static int Main(string[] args)
+        {
+            int retValue = 0;
+
+            bool mode = bool.Parse(args[0]);
+            if (mode)
+            {
+                if (Thread.CurrentThread.GetApartmentState() == ApartmentState.MTA)
+                {
+                    retValue = 1;
```
Comment (danmosemsft) [inline src/System.Threading.Thread/tests/MTAMain/MTAMain.cs]: use zero for success, non zero for failure in console apps

File: src/System.Threading.Thread/tests/ThreadTests.cs [hunk 11]
```diff
@@ -139,6 +144,58 @@ yield return
                 };
         }
 
+        [Theory]
+        [InlineData(true)] 
+        [InlineData(false)]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+        public static void ApartmentState_STAAtributePresent(bool mode)
+        {
+            const string AppName = "STAMain.exe";
```
Comment (danmosemsft) [inline src/System.Threading.Thread/tests/ThreadTests.cs]: Duplicated code. Don't duplicate code if you can avoid it. Minimise each test

File: src/System.Threading.Thread/tests/UnknownMain/UnknownMain.cs [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        static int Main(string[] args)
+        {
+            int retValue = 0;
+
+            int mode = int.Parse(args[0]);
+            if (mode == 1)
+            {
+                if (Thread.CurrentThread.GetApartmentState() == ApartmentState.Unknown)
+                {
+                    retValue = 1;
```
Comment (danmosemsft) [inline src/System.Threading.Thread/tests/UnknownMain/UnknownMain.cs]: as mentioned in the other review, I think we want this to remain MTA

File: src/System.Threading.Thread/System.Threading.Thread.sln [hunk 13]
```diff
@@ -20,24 +20,68 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{E107E9C1-E89
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "ref", "ref", "{2E666815-2EDB-464B-9DF6-380BF4789AD4}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "UnknownMain", "tests\UnknownMain\UnknownMain.csproj", "{C8E0B248-31E9-4175-B3B0-7729193227FB}"
```
Comment (weshaggard) [inline src/System.Threading.Thread/System.Threading.Thread.sln]: Can you please undo the changes to this solution and use the `/t:UpdateVSConfigurations` target to generate the solution.

File: src/System.Threading.Thread/tests/UnknownMain/UnknownMain.cs [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+namespace Unknown
+{
+    internal static class Program
+    {
+        static int Main(string[] args)
+        {
+            int retValue = 0;
+
+            int mode = int.Parse(args[0]);
+            if (mode == 1)
```
Comment (danmosemsft) [inline src/System.Threading.Thread/tests/UnknownMain/UnknownMain.cs]: I think a string would be clearer than an int for the mode because then I don't have to check both sides to see it's correct

File: src/System.Threading.Thread/tests/UnknownMain/UnknownMain.runtimeconfig.json [hunk 15]
```diff
@@ -0,0 +1,8 @@
+{
+	"runtimeOptions": {
```
Comment (danmosemsft) [inline src/System.Threading.Thread/tests/UnknownMain/UnknownMain.runtimeconfig.json]: spacing

File: src/System.Threading.Thread/tests/ThreadTests.cs [hunk 16]
```diff
@@ -139,6 +144,24 @@ yield return
                 };
         }
 
+        [Theory]
+        [InlineData("STAMain.exe", "GetApartmentState")] 
+        [InlineData("STAMain.exe", "SetApartmentState")]
+        [InlineData("MTAMain.exe", "GetApartmentState")]
+        [InlineData("MTAMain.exe", "SetApartmentState")]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+        public static void ApartmentState_AtributePresent(string AppName, string mode)
```
Comment (danmosemsft) [inline src/System.Threading.Thread/tests/ThreadTests.cs]: good use of inlinedata!

**Comment by Anipik** [comment]
@danmosemsft  tests were already there for secondary threads. I added tests for noAttributeCase

**Comment by Anipik** [comment]
@dotnet-bot help

**Comment by Anipik** [comment]
@dotnet-bot test this please

File: src/System.Threading.Thread/tests/MTAMain/MTAMain.exe.config (hunk 7 above)
Comment (Anipik) [reply src/System.Threading.Thread/tests/MTAMain/MTAMain.exe.config]: This file is required to run the tests for netframework

**Comment by Anipik** [comment]
@jkotas what is the behavior of STAThread attribute, SAS and GAS on windows nano server ?

File: src/System.Threading.Thread/tests/MTAMain/MTAMain.exe.config (hunk 7 above)
Comment (safern) [reply src/System.Threading.Thread/tests/MTAMain/MTAMain.exe.config]: We need this .exe.config files in order to run the tests against the generated runtime folder by the repo built because without this it would try to run them vs the GAC.

**Comment by Anipik** [comment]
@dotnet-bot test Linux x64 Release Build

**Comment by Anipik** [comment]
@dotnet-bot test Linux x64 Release Build

**Comment by jkotas** [comment]
>  what is the behavior of STAThread attribute, SAS and GAS on windows nano server ?

It should work as expected, I think. Do you see it failing?

**Comment by danmosemsft** [comment]
Possibly CoInitializeEx is a stub on Nano, like ShellExecEx is...?

**Comment by Anipik** [comment]
@jkotas  I guess it is initializing it to MTA

**Comment by Anipik** [comment]
@dotnet-bot test Windows x64 Debug Build

**Comment by Anipik** [comment]
@dotnet-bot Linux x64 Release Build

**Comment by Anipik** [comment]
Tests were failing for STA Earlier

**Comment by Anipik** [comment]
@dotnet-bot Linux x64 Release Build

**Comment by Anipik** [comment]
@dotnet-bot test this please

**Comment by Anipik** [comment]
@dotnet-bot test Linux x64 Release Build

**Comment by Anipik** [comment]
@jkotas @danmosemsft can you take a last look before I merge ?

**Comment by jkotas** [comment]
> @pjanotti should review if he's in, or else maybe @ianhays, as IO devs past and present

@danmosemsft I think that you meant this comment  for the GetFullPath PRs. (Some commits from the GetFullPath PR got mixed into this one, and then reverted.)

**Comment by jkotas** [comment]
@Anipik Could you please squash the commits into one, and write a good description for it?

**Comment by danmosemsft** [comment]
@jkotas ah, thanks.

**Comment by jkotas** [comment]
Marking as blocked - the matching CoreCLR change is getting reverted: https://github.com/dotnet/coreclr/pull/15586

**Comment by Anipik** [comment]
There was an error while rebasing as I have updated my master. I have cherry picked the squash changed and created a new PR
https://github.com/dotnet/corefx/pull/26001
